### DISCUSSION TITLE
Problems with measurement

### DISCUSSION THREAD

**Discussion by YTheaudin96t** [geoman-io/leaflet-geoman#1347] [Questions & Help] [answered]
Hi,

I have two 2 problems

1 > showTooltipOnHover: true  not working 

My code `map.pm.setGlobalOptions({ measurements: { measurement: true, displayFormat: 'metric', showTooltipOnHover: true, showTooltip: true, radius: true, area: true, segmentLength: true, totalLength: true, showTooltip: true, perimeter: true, coordinates: true } })`

Resize circle OK
![image](https://user-images.githubusercontent.com/96052160/233652220-b94d8b1d-e906-4bf4-97ef-020bcc3ec77e.png)

Mouse hover KO
![image](https://user-images.githubusercontent.com/96052160/233652644-94d77b1f-2415-4388-be2a-e805ddf79856.png)

> 2
I would like to retrieve element select , + its measure and create a makerText do you have a solution

Regards,
Yvan

**Comment by YTheaudin96t** [comment]
Complément by contextMenu

**Comment by Falke-Design** [comment]
The first issue is a real bug (#1348), we will fix this in the next release. Until then you can use this workaround:
```
map.on('pm:create',(x)=>{
  x.layer.on('pm:edit',(e)=>{
    e.layer.pm.calcMeasurement(e.layer);
    e.layer.pm.addMeasurementTooltipToLayer(e.layer);
  })
})
```

Sorry but I don't understand your second point, can you please try to explain it more

**Reply by YTheaudin96t** [reply]
I want to retrieve the measurements and insert them into a MarkerText from a contextMenu

**Reply by YTheaudin96t** [reply]
I'm sorry on the current map it works after backup and restore it doesn't work anymore

**Reply by Falke-Design** [reply]
You need to apply the workaround on the imported layers too. For example after importing with `L.GeoJSON()`:
```
var layers = L.geoJSON(YOUR_GEOJSON);
layer.eachLayer((layer)=>{
  layer.on('pm:edit',(e)=>{
    e.layer.pm.calcMeasurement(e.layer);
    e.layer.pm.addMeasurementTooltipToLayer(e.layer);
  })
})
```

**Reply by Falke-Design** [reply]
What do you mean with `MarkerText from a contextMenu`? Do you mean a `TextMarker`? What has this to do with a contextmenu?

**Reply by YTheaudin96t** [reply]
Yes sorry

**Reply by YTheaudin96t** [reply]
![image](https://user-images.githubusercontent.com/96052160/233671189-d18d266b-635d-440e-8128-75076bb7b61c.png)
with the Add Size I want to insert the measurements
sorry if i'm not clear

**Reply by YTheaudin96t** [reply]
> You need to apply the workaround on the imported layers too. For example after importing with `L.GeoJSON()`:
> 
> ```
> var layers = L.geoJSON(YOUR_GEOJSON);
> layer.eachLayer((layer)=>{
>   layer.on('pm:edit',(e)=>{
>     e.layer.pm.calcMeasurement(e.layer);
>     e.layer.pm.addMeasurementTooltipToLayer(e.layer);
>   })
> })
> ```

I always try the same thing no tooltips

  function createLayersFromJson(data) {
             const layers = [];
             var layerGroup = L.layerGroup();
             
             L.geoJSON(data, {
                     pointToLayer: (feature, latlng) => {
                     //console.log(feature.properties);
                     if (feature.properties.radius) {
                         //console.log(feature);
                             layerGroup.addLayer( new L.Circle(latlng, feature.properties.radius));
                     } else {
                         if (feature.properties.TextInfo) {
                             layerGroup.addLayer(new L.Marker(latlng, {
                                 textMarker: true,
                                 text: feature.properties.TextInfo}));
                         }
                         else if (feature.properties.iconUrl =='Telepilote') {
                             layerGroup.addLayer(new L.Marker(latlng, { icon: Telepilote }));
                         }
                         else if (feature.properties.iconUrl =='Spotter') {
                             layerGroup.addLayer(new L.Marker(latlng, { icon: Spotter }));
                         }
                         else {
                             
                             layerGroup.addLayer(new L.Marker(latlng));
                         }
                             
                         }
                     },
                     onEachFeature: (feature, layer) => {
                         
                         layer.on('pm:edit', (e) => {
                             e.layer.pm.calcMeasurement(e.layer);
                             e.layer.pm.addMeasurementTooltipToLayer(e.layer);
                         })
                         layerGroup.addLayer(layer)
                     },
                 });

             return layerGroup;
         };

**Reply by Falke-Design** [reply, marked as answer]
I know now the problem. The layer has after importing no measurement and also the GlobalOptions (`showTooltipOnHover`) are not applied. Use this for now:

```
var layers = L.geoJSON(JSON.parse(data)).addTo(map);
map.pm.setGlobalOptions({ measurements: { measurement: true, displayFormat: 'metric', showTooltipOnHover: true, showTooltip: true} })
layers.eachLayer((layer)=>{
  layer.on('pm:edit',(e)=>{
    e.layer.pm.calcMeasurement(e.layer);
    e.layer.pm.addMeasurementTooltipToLayer(e.layer);
  });

  layer.pm.calcMeasurement(layer);
  layer.pm.addMeasurementTooltipToLayer(layer);
  layer.closeTooltip();
})
```

PS: It is important to apply the measurement logic exactly like above and **not** in `onEachFeature`.

I try to fix this with the next version too.

**Reply by Falke-Design** [reply]
> ![image](https://user-images.githubusercontent.com/96052160/233671189-d18d266b-635d-440e-8128-75076bb7b61c.png) 
with the Add Size I want to insert the measurements sorry if i'm not clear

```
function onSizeBtnClick(layer){
  const measurements = layer.pm.measurements || {};

  const text = Object.entries(measurements).map(([key, value]) => `${key}: ${value}`).join('\n');

  let latlng;

  if(layer.getLatLng){
    latlng = layer.getLatLng();
  } else if(layer.getCenter) {
    latlng = layer.getCenter()
  } else if(layer.getBounds) {
    latlng = layer.getBounds().getCenter();
  } else if(layer.getBounds) {
    latlng = layer.getBounds().getCenter();
  } else if(layer.getLatLngs){
    latlng = L.latLngBounds(layer.getLatLngs()).getCenter();
  } else {
    throw new Error('Center not found!');
  }

  L.marker(latlng, {
    textMarker: true,
    text: text,
  }).addTo(map);
}
```
![grafik](https://user-images.githubusercontent.com/19800037/233682511-b2efb790-385f-4ac5-b09a-1d68e1e39371.png)

**Reply by YTheaudin96t** [reply]
Thank you for the answer,

1. on the other hand how you select the layer

sorry i start with your library
